rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+than just outputting a prompt and waiting until it's completed, it can send a
+notification back to the dependent client about the prompt. In some cases, like
+a touch prompt, the dependent client can mirror the agent's prompt:
+
+```console
+> tsh ssh server01
+### sign through agent, agent notifies client of touch prompt requirement
+Tap your YubiKey
+```
+
+In other cases, like pin entry, the dependent client can continue to output the
+redirection prompt with additional context:
+
+```console
+> tsh ssh server01
+Go to your Teleport Key Agent and enter your PIV PIN to continue.
+```
+
+##### Confirmation prompt (*future work*)
+
+We can add a confirmation layer between the agent server and dependent clients.
+In order to utilize the agent to query certs or sign with keys, the user would
+need to confirm the dependent client's connection with one or more of the following:
+
+* basic confirmation (y/n) prompt in the key agent, similar to what we use for headless
+* password prompt in the dependent client
+  * like with ssh-agent, this would be a temporary password provided on agent
+  startup
+* touch prompt with temporary local MFA registration
+
+Once the connection is authenticated through the local confirmation mechanism,
+the dependent client could perform any queries or signatures until client is
+closed. In practice, this means that each `tsh` command would require a single
+confirmation prompt to complete.
+
+Note: once again, we run into the issue of prompting on every `tsh` command
+that we ran into with `hardware_key_pin`. Therefore we would only want to
+enable this confirmation layer when the tradeoff of better security for worse
+UX is justified. See the sections on [the risk of key agent forwarding](#key-agent-forwarding)
+and [safely supporting key agent forwarding](#support-teleport-key-agent-forwarding-future-work)
+for more details on how and why this confirmation layer would be included.
+
+### Modes
+
+The Teleport Key Agent can work in two different modes; `signing` and `login sync`.
+
+Note: the Teleport Key Agent will be pre-released with `signing` mode only as
+it will greatly improve the UX for Hardware Key pin support, especially when
+paired with [Hardware Key pin caching](./0198-hardware-key-pin-caching).
+
+#### Signing mode
+
+In signing mode, the Teleport Key Agent will only be responsible for providing
+a signing interface for Teleport Clients, similar to ssh-agent or gpg-agent.
+All other login information (profile info, certificates, etc.) will be retrieved
+from the standard file storage interface.
+
+This signing mode provides the baseline feature set needed to introduce UX
+improvements for features which limit access to private keys, including:
+
+* Hardware Key Support
+* [Memory Key Storage (*future work*)](#memory-key-storage)
+
+Since only private keys are synced between clients in this mode, there is a
+possibility of state drift for other login information.
```

Comment:
Should the agent be aware of the distinct "sessions"?

Reply:
Can you elaborate?